apps/web/app/api/auth/[...nextauth]/authOptions.ts

```diff
@@ -57,6 +60,62 @@ export const authOptions: NextAuthOptions = {
           throw new Error("No user matches the provided credentials");
         }
 
+        if (user.twoFactorEnabled && credentials.backupCode) {
+          if (!process.env.FORMBRICKS_ENCRYPTION_KEY) {
+            console.error("Missing encryption key; cannot proceed with backup code login.");
+            throw new Error("Internal Server Error");
+          }
+
+          if (!user.backupCodes) throw new Error("No backup codes found");
+
+          const backupCodes = JSON.parse(
+            symmetricDecrypt(user.backupCodes, process.env.FORMBRICKS_ENCRYPTION_KEY)
+          );
+
+          // check if user-supplied code matches one
+          const index = backupCodes.indexOf(credentials.backupCode.replaceAll("-", ""));
+          if (index === -1) throw new Error("Invalid backup code");
+
+          // delete verified backup code and re-encrypt remaining
+          backupCodes[index] = null;
+          await prisma.user.update({
+            where: {
+              id: user.id,
+            },
+            data: {
+              backupCodes: symmetricEncrypt(
+                JSON.stringify(backupCodes),
+                process.env.FORMBRICKS_ENCRYPTION_KEY
+              ),
+            },
+          });
+        } else if (user.twoFactorEnabled) {
+          if (!credentials.totpCode) {
+            throw new Error("second factor required");
+          }
+
+          if (!user.twoFactorSecret) {
+            throw new Error("Internal Server Error");
+          }
+
+          if (!process.env.FORMBRICKS_ENCRYPTION_KEY) {
+            throw new Error("Internal Server Error");
+          }
+
+          const secret = symmetricDecrypt(user.twoFactorSecret, process.env.FORMBRICKS_ENCRYPTION_KEY);
+          if (secret.length !== 32) {
```

Comment:
We also introduced the `FORMBRICKS_ENCRYPTION_KEY` with single use surveys. Here it's setup for length 24. Is this also working or should we update the length of the encryption key to 32?